File: 백준/Silver/3085. 사탕 게임/사탕 게임.js
const input = require('fs')
  .readFileSync('/dev/stdin')
  .toString()
  .trim()
  .split('\n');

const n = Number(input[0]);
const board = input.slice(1, 1 + n).map((row) => row.split(''));
let answer = 0;

// 바뀐 보드의 사탕 갯수 확인하기
function check(board) {
  // 행 확인

  for (let x = 0; x < n; x++) {
    let tmp = 1;
    for (let y = 0; y < n - 1; y++) {
      if (board[x][y] === board[x][y + 1]) {
        tmp += 1;
        answer = Math.max(answer, tmp);
      } else {
        tmp = 1;
      }
    }
  }

  for (let y = 0; y < n; y++) {
    let tmp = 1;
    for (let x = 0; x < n - 1; x++) {
      if (board[x][y] === board[x + 1][y]) {
        tmp += 1;
        answer = Math.max(answer, tmp);
      } else {
        tmp = 1;
      }
    }
  }
}

const dx = [0, 1]; // 동 남
const dy = [1, 0]; // 동 남

for (let x = 0; x < n; x++) {
  for (let y = 0; y < n; y++) {
    for (let idx = 0; idx < 2; idx++) {
      const nx = x + dx[idx];
      const ny = y + dy[idx];

      if (nx >= n || ny >= n) continue; // board 안에만 접근하도록 예외 처리
      let tempCandy = board[x][y];
      board[x][y] = board[nx][ny];
      board[nx][ny] = tempCandy;
      // 사탕 바꾸기
      check(board); // 바뀐 사탕으로 행과 열 확인하기
      tempCandy = board[x][y];
      board[x][y] = board[nx][ny];
      board[nx][ny] = tempCandy;
      // 다시 사탕 바꾸기 (원본 보드 유지)
    }
  }
}
console.log(answer);

File: 백준/Silver/3085. 사탕 게임/사탕 게임.py
import sys
input=sys.stdin.readline

def check(arr):
    n=len(arr)
    answer=1

    for i in range(n):
        # 열 순회하면서 연속되는 숫자 세기
        cnt=1
        for j in range(1, n):
            if arr[i][j] == arr[i][j-1]:
        	# 이전 것과 같다면 cnt에 1 더하기
                cnt += 1
            else:
            # 이전과 다르다면 다시 1로 초기화
                cnt=1
                
            # 비교해서 현재 cnt가 더 크다면 answer 갱신하기
            if cnt > answer:
                answer = cnt

        # 행 순회하면서 연속되는 숫자 세기
        cnt=1
        for j in range(1, n):
            if arr[j][i] == arr[j-1][i]:
        	# 이전 것과 같다면 cnt에 1 더하기
                cnt += 1
            else:
            # 이전과 다르다면 다시 1로 초기화
                cnt=1
                
            # 비교해서 현재 cnt가 더 크다면 answer 갱신하기
            if cnt > answer:
                answer = cnt

    return answer

n=int(input())
arr=[list(input()) for _ in range(n)]
answer=0

for i in range(n):
    for j in range(n):
        # 열 바꾸기
        if j+1 < n:
        	# 인점한 것과 바꾸기
            arr[i][j], arr[i][j+1] = arr[i][j+1], arr[i][j]
            
            # check는 arrd에서 인점한 것과 바꿨을 때 가장 긴 연속한 부분을 찾아내는 함수이다
            temp=check(arr)

            if temp > answer:
                answer = temp
               
            # 바꿨던 것을 다시 원래대로 돌려놓기
            arr[i][j], arr[i][j+1] = arr[i][j+1], arr[i][j]

        # 행 바꾸기
        if i+1 < n:
        	# 인점한 것과 바꾸기
            arr[i][j], arr[i+1][j] = arr[i+1][j], arr[i][j]
            
            # check는 arrd에서 인점한 것과 바꿨을 때 가장 긴 연속한 부분을 찾아내는 함수이다
            temp=check(arr)

            if temp > answer:
                answer = temp
            
            # 바꿨던 것을 다시 원래대로 돌려놓기
            arr[i][j], arr[i+1][j] = arr[i+1][j], arr[i][j]
            
print(answer)